### PR TITLE
S3UTILS-238: regression test for crrExistingObjects with non-ASCII keys

### PR DESCRIPTION
**Pull request by tcarmet**
## Intent: why does this change exist?

S3C-11235 / RD-1751 reported that `crrExistingObjects` fails with `SignatureDoesNotMatch` (HTTP 403) for any object whose key contains non-ASCII characters such as Polish diacritics, blocking the customer from replicating ~131k existing objects. The underlying bug was fixed in [`@scality/cloudserverclient` 1.0.8](https://scality.atlassian.net/browse/CLDSRVCLT-13), which s3utils picked up in [S3UTILS-232](https://scality.atlassian.net/browse/S3UTILS-232), but no automated test exercises the path. This change adds the regression test so any future encoding regression in the dependency stack is caught in CI.

**References:** [S3UTILS-238](https://scality.atlassian.net/browse/S3UTILS-238) · [S3C-11235](https://scality.atlassian.net/browse/S3C-11235) · [RD-1751](https://scality.atlassian.net/browse/RD-1751) · [CLDSRVCLT-13](https://scality.atlassian.net/browse/CLDSRVCLT-13) · [S3UTILS-232](https://scality.atlassian.net/browse/S3UTILS-232)

[S3UTILS-232]: https://scality.atlassian.net/browse/S3UTILS-232?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[S3UTILS-238]: https://scality.atlassian.net/browse/S3UTILS-238?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[S3C-11235]: https://scality.atlassian.net/browse/S3C-11235?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RD-1751]: https://scality.atlassian.net/browse/RD-1751?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CLDSRVCLT-13]: https://scality.atlassian.net/browse/CLDSRVCLT-13?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/functional/crrExistingObjects.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        for (const obj of testObjects) {
+            const head = await accountSource.s3Client.send(new HeadObjectCommand({
+                Bucket: accountSource.bucketName,
+                Key: obj.Key,
+            }));
+            expect(head.ReplicationStatus).toBe('PENDING');
+        }
+    }
+
+    it('should replicate existing objects whose key contains Polish diacritics', async () => {
```
Comment (BourgoisMickael) [inline tests/functional/crrExistingObjects.js:122]: maybe we should also test all the utf8 unicode planes, with the 3 byte range and 4 bytes range as well. I think most of them are acceptable in s3 object key

**Review by BourgoisMickael — APPROVED** [review]

**Review by BourgoisMickael — APPROVED** [review]

File: tests/utils/S3Setup.js [hunk 2]
```diff
@@ -202,9 +202,163 @@ async function deleteTestAccount(vaultClient, account) {
...
+        VersioningConfiguration: { Status: 'Enabled' },
+    }));
+
+    log.info('Creating IAM policies and roles for CRR');
+    const policy = {
+        Version: '2012-10-17',
+        Statement: [
+            {
+                Effect: 'Allow',
+                Action: ['s3:GetObjectVersion', 's3:GetObjectVersionAcl', 's3:ReplicateObject'],
```
Comment (jonathan-gramain) [inline tests/utils/S3Setup.js]: style nitpick

```suggestion
                Action: [
                    's3:GetObjectVersion',
                    's3:GetObjectVersionAcl',
                    's3:ReplicateObject',
                ],
```

File: tests/functional/crrExistingObjects.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            warn: wrap('warn'),
+            error: wrap('error'),
+            fatal: wrap('fatal'),
+            debug: wrap('debug'),
+            trace: wrap('trace'),
+        },
+    };
+}
+
+function runUpdater(updater) {
```
Comment (jonathan-gramain) [inline tests/functional/crrExistingObjects.js]: I think this can be replaced at the call site by `await promisify(updater.run)`

File: tests/functional/crrExistingObjects.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+const log = new Logger('crrExistingObjects:test');
+const SITE_NAME = 'test-site';
+
+function makeCaptureLogger() {
+    const captured = [];
+    const inner = new Logger('crrExistingObjects:test:capture');
+    const wrap = level => (msg, data) => {
+        captured.push({ level, message: msg, ...(data || {}) });
+        return inner[level](msg, data);
+    };
```
Comment (jonathan-gramain) [inline tests/functional/crrExistingObjects.js]: Adding parenthesis would show more clearly what is returned by `wrap`

```suggestion
    const wrap = level => (
        (msg, data) => {
            captured.push({ level, message: msg, ...(data || {}) });
            return inner[level](msg, data);
        }
    );
```

File: tests/functional/crrExistingObjects.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    afterEach(async () => {
+        log.info('Cleaning up test accounts');
+        await removeCrrConfiguration(accountSource);
+        await removeCrrConfiguration(accountDest);
+        await deleteTestAccount(vaultClient, accountSource);
+        await deleteTestAccount(vaultClient, accountDest);
+        log.info('Test accounts deleted');
+    });
+
+    async function runAndAssert(testObjects) {
```
Comment (jonathan-gramain) [inline tests/functional/crrExistingObjects.js]: Nitpick: the function is called `runAndAssert` but there is no call to `assert`, just `expect`, what about simply naming it `runTest`

**Review by jonathan-gramain — APPROVED** [review]

**Review by jonathan-gramain — APPROVED** [review]
